File: backend/src/models/property.py
"""Property and PropertyAssignment models.

Implements T025 and T026 from tasks.md.
"""
from datetime import datetime
from typing import Optional
from uuid import uuid4

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from ..core.database import Base


class Property(Base):
    """Property model representing a rental property.
    
    Properties are owned by users with role='owner' and managed by
    intermediaries through PropertyAssignment junction table.
    
    Features:
    - Multi-unit support (apartments, condos, etc.)
    - Immutable base currency for financial consistency
    - Full address tracking
    - Owner and intermediary access control via RLS
    """

    __tablename__ = "properties"

    # Primary Key
    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
        index=True,
        comment="Unique property identifier",
    )

    # Ownership
    owner_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Property owner (must be role='owner')",
    )

    # Basic Info
    name = Column(
        String(255),
        nullable=False,
        comment="Property name/identifier",
    )

    # Address
    address_line1 = Column(
        String(255),
        nullable=False,
        comment="Street address",
    )
    address_line2 = Column(
        String(255),
        nullable=True,
        comment="Apartment/unit number",
    )
    city = Column(
        String(100),
        nullable=False,
        comment="City",
    )
    state = Column(
        String(100),
        nullable=False,
        comment="State/province",
    )
    postal_code = Column(
        String(20),
        nullable=False,
        comment="Postal/ZIP code",
    )
    country = Column(
        String(100),
        nullable=False,
        comment="Country",
    )

    # Property Details
    total_units = Column(
        Integer,
        nullable=False,
        comment="Number of rental units",
    )
    base_currency = Column(
        String(3),
        nullable=False,
        comment="ISO 4217 currency code (immutable after creation)",
    )

    # Timestamps
    created_at = Column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
        comment="Property creation timestamp",
    )
    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
        comment="Last update timestamp",
    )

    # Relationships
    owner = relationship(
        "User",
        back_populates="owned_properties",
        foreign_keys=[owner_id],
    )
    assignments = relationship(
        "PropertyAssignment",
        back_populates="property",
        cascade="all, delete-orphan",
    )
    tenants = relationship(
        "Tenant",
        back_populates="property",
        cascade="all, delete-orphan",
    )
    payments = relationship(
        "Payment",
        back_populates="property",
        cascade="all, delete-orphan",
    )

    # Constraints
    __table_args__ = (
        CheckConstraint("total_units > 0", name="check_total_units_positive"),
        Index("idx_properties_city", "city", "state"),
    )

    def __repr__(self) -> str:
        return f"<Property(id={self.id}, name={self.name}, owner_id={self.owner_id})>"

    @property
    def full_address(self) -> str:
        """Get formatted full address."""
        parts = [
            self.address_line1,
            self.address_line2,
            self.city,
            self.state,
            self.postal_code,
            self.country,
        ]
        return ", ".join(filter(None, parts))


class PropertyAssignment(Base):
    """PropertyAssignment junction table for intermediary assignments.
    
    Manages many-to-many relationship between intermediaries and properties.
    Allows owners to assign multiple intermediaries to a property and
    intermediaries to manage multiple properties.
    
    Features:
    - Soft delete (removed_at) for assignment history
    - Unique constraint on active assignments
    - Audit trail (assigned_by, assigned_at, removed_at)
    """

    __tablename__ = "property_assignments"

    # Primary Key
    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
        index=True,
        comment="Assignment identifier",
    )

    # Foreign Keys
    property_id = Column(
        UUID(as_uuid=True),
        ForeignKey("properties.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Assigned property",
    )
    intermediary_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Assigned intermediary (must be role='intermediary')",
    )
    assigned_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=False,
        comment="Owner who made assignment",
    )

    # Status
    is_active = Column(
        Boolean,
        default=True,
        nullable=False,
        comment="Active assignment flag",
    )

    # Timestamps
    assigned_at = Column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
        comment="Assignment timestamp",
    )
    removed_at = Column(
        DateTime,
        nullable=True,
        comment="Removal timestamp (soft delete)",
    )

    # Relationships
    property = relationship(
        "Property",
        back_populates="assignments",
        foreign_keys=[property_id],
    )
    intermediary = relationship(
        "User",
        back_populates="property_assignments",
        foreign_keys=[intermediary_id],
    )
    assigner = relationship(
        "User",
        foreign_keys=[assigned_by],
    )

    # Constraints
    __table_args__ = (
        UniqueConstraint(
            "property_id",
            "intermediary_id",
            "is_active",
            name="uq_property_intermediary_active",
            postgresql_where=(is_active == True),
        ),
        Index("idx_pa_active", "is_active", postgresql_where=(is_active == True)),
    )

    def __repr__(self) -> str:
        return f"<PropertyAssignment(id={self.id}, property_id={self.property_id}, intermediary_id={self.intermediary_id})>"

File: mobile/lib/models/property.dart
/// Property model for local storage and API communication.
///
/// Implements T041 from tasks.md.
library;

/// Property model with all fields from backend.
class Property {
  Property({
    required this.id,
    required this.ownerId,
    required this.name,
    required this.addressLine1,
    this.addressLine2,
    required this.city,
    required this.state,
    required this.postalCode,
    required this.country,
    required this.totalUnits,
    required this.baseCurrency,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Create Property from JSON (API response).
  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['id'] as String,
      ownerId: json['owner_id'] as String,
      name: json['name'] as String,
      addressLine1: json['address_line1'] as String,
      addressLine2: json['address_line2'] as String?,
      city: json['city'] as String,
      state: json['state'] as String,
      postalCode: json['postal_code'] as String,
      country: json['country'] as String,
      totalUnits: json['total_units'] as int,
      baseCurrency: json['base_currency'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  /// Create Property from SQLite database row.
  factory Property.fromMap(Map<String, dynamic> map) {
    return Property(
      id: map['id'] as String,
      ownerId: map['owner_id'] as String,
      name: map['name'] as String,
      addressLine1: map['address_line1'] as String,
      addressLine2: map['address_line2'] as String?,
      city: map['city'] as String,
      state: map['state'] as String,
      postalCode: map['postal_code'] as String,
      country: map['country'] as String,
      totalUnits: map['total_units'] as int,
      baseCurrency: map['base_currency'] as String,
      createdAt: DateTime.parse(map['created_at'] as String),
      updatedAt: DateTime.parse(map['updated_at'] as String),
    );
  }
  final String id;
  final String ownerId;
  final String name;
  final String addressLine1;
  final String? addressLine2;
  final String city;
  final String state;
  final String postalCode;
  final String country;
  final int totalUnits;
  final String baseCurrency;
  final DateTime createdAt;
  final DateTime updatedAt;

  /// Full address as a single string.
  String get fullAddress {
    final parts = [
      addressLine1,
      if (addressLine2 != null && addressLine2!.isNotEmpty) addressLine2,
      city,
      state,
      postalCode,
      country,
    ];
    return parts.join(', ');
  }

  /// Convert Property to JSON for API requests.
  Map<String, dynamic> toJson() => {
        'id': id,
        'owner_id': ownerId,
        'name': name,
        'address_line1': addressLine1,
        'address_line2': addressLine2,
        'city': city,
        'state': state,
        'postal_code': postalCode,
        'country': country,
        'total_units': totalUnits,
        'base_currency': baseCurrency,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt.toIso8601String(),
      };

  /// Convert Property to SQLite database row.
  Map<String, dynamic> toMap() => {
        'id': id,
        'owner_id': ownerId,
        'name': name,
        'address_line1': addressLine1,
        'address_line2': addressLine2,
        'city': city,
        'state': state,
        'postal_code': postalCode,
        'country': country,
        'total_units': totalUnits,
        'base_currency': baseCurrency,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt.toIso8601String(),
      };

  /// Create a copy with updated fields.
  Property copyWith({
    String? id,
    String? ownerId,
    String? name,
    String? addressLine1,
    String? addressLine2,
    String? city,
    String? state,
    String? postalCode,
    String? country,
    int? totalUnits,
    String? baseCurrency,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) =>
      Property(
        id: id ?? this.id,
        ownerId: ownerId ?? this.ownerId,
        name: name ?? this.name,
        addressLine1: addressLine1 ?? this.addressLine1,
        addressLine2: addressLine2 ?? this.addressLine2,
        city: city ?? this.city,
        state: state ?? this.state,
        postalCode: postalCode ?? this.postalCode,
        country: country ?? this.country,
        totalUnits: totalUnits ?? this.totalUnits,
        baseCurrency: baseCurrency ?? this.baseCurrency,
        createdAt: createdAt ?? this.createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );

  @override
  String toString() =>
      'Property(id: $id, name: $name, city: $city, totalUnits: $totalUnits)';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Property && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}
